### PR TITLE
Bug/692 badge view

### PR DESCRIPTION
**Pull request by jmmortega**
### Description of Change ###

Improve round BadgeView. (A better approach maybe is measure Label text to get size an create round badge from here) 

### Bugs Fixed ###
#692 and #641 

### API Changes ###

Added: 
 
- A new BindableProperty for BadgeView called IsRounded to use feature to round Badges
- Include a new sample inside of sample app, to show how working IsRounded

Changed:

 - Checks to get BorderRadius to get a better rounded effect.
 - I maintain legacy way to create border in BadgeViews
 

### Behavioral Changes ###

Have a new property inside of BadgeView that for default is True so, in this case devs must be check if BadgeView adapt to design.

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [ X ] Has samples (if omitted, state reason in description)
- [ ] Rebased on top of main at time of PR
- [ X ] Changes adhere to coding standard

- [ ] Updated [documentation](https://github.com/MicrosoftDocs/xamarin-communitytoolkit)

### CODE REVIEW

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 1]
```diff
@@ -157,6 +159,25 @@ public FontAttributes FontAttributes
...
+        
+
+
+        public static BindableProperty IsRoundedProperty =
+	        BindableProperty.Create(nameof(IsRounded), typeof(bool), typeof(BadgeView), true,
+		        propertyChanged:IsRoundedChanged);
+
+        
+
+
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:174]:
```suggestion

```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 2]
```diff
@@ -157,6 +159,25 @@ public FontAttributes FontAttributes
             set => SetValue(FontAttributesProperty, value);
         }
 
+        static void IsRoundedChanged(BindableObject bindable, object oldvalue, object newvalue) =>
+	        (bindable as BadgeView)?.UpdateRounded();
+
+        
+
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:166]:
```suggestion
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 3]
```diff
@@ -157,6 +159,18 @@ public FontAttributes FontAttributes
             set => SetValue(FontAttributesProperty, value);
         }
 
+        static void IsRoundedChanged(BindableObject bindable, object oldvalue, object newvalue) =>
+	        (bindable as BadgeView)?.UpdateRounded();
+
+        public static BindableProperty IsRoundedProperty =
+	        BindableProperty.Create(nameof(IsRounded), typeof(bool), typeof(BadgeView), true,
+		        propertyChanged:IsRoundedChanged);
```
Comment (AndreiMisiukevich) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:167]:
```suggestion
		        propertyChanged:IsRoundedChanged);
		        
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 4]
```diff
@@ -279,11 +306,19 @@ void UpdateBadgeViewPlacement(bool force = false)
...
             }
 
+            BadgeText.Margin = GetMarginsLengthIsOne();
             BadgeIndicatorContainer.Margin = containerMargin;
             BadgeContent.Margin = contentMargin;
             placementDone = true;
         }
 
+        private Thickness GetMarginsLengthIsOne()
+			=> BadgeText.Text.Length == 1 && IsRounded
```
Comment (Cfun1) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:310]: isn't more readable with a nested if else statement instead ?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 5]
```diff
@@ -157,6 +159,25 @@ public FontAttributes FontAttributes
             set => SetValue(FontAttributesProperty, value);
         }
 
+        static void IsRoundedChanged(BindableObject bindable, object oldvalue, object newvalue) =>
```
Comment (Cfun1) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:162]: I suggest to follow the same order creation of BindableProperty, getter/setter, property changed event:
````
BindableProperty.Create
...
public bool IsRounded
...
static void IsRoundedChanged(BindableObject bindable, object oldvalue, object newvalue) => 
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 6]
```diff
@@ -279,11 +306,19 @@ void UpdateBadgeViewPlacement(bool force = false)
...
             BadgeContent.Margin = contentMargin;
             placementDone = true;
         }
 
+        private Thickness GetMarginsLengthIsOne()
+			=> BadgeText.Text.Length == 1 && IsRounded
+				? Device.RuntimePlatform == Device.Android  
+					? new Thickness(10,0,10,0)
+					: new Thickness(8,0,8,0)
+				: new Thickness(4,0,4,0);
```
Comment (Cfun1) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:314]: isn't more readable with a nested if else statement instead ?

Comment (pictos) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:314]: About these `magic numbers`, will they work in all screen resolutions/sizes?

**Comment by jfversluis** [comment]
@jmmortega are you able to still answer the questions and/or make the requested changes?

If not, no problem! Just let us know so we can take over :)

Thanks!

**Comment by jmmortega** [comment]
@jfversluis Yep!. Sorry! This weekend complete all your requests!

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs (hunk 6 above)
Comment (jmmortega) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:314]: @pictos Magic numbers working well, testing in several resolutions, sizes and badgeview seems corectly

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs (hunk 5 above)
Comment (jmmortega) [reply src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs]: Changed!

**Comment by Cfun1** [comment]
@jmmortega Could you please resolve the conflict ?

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 7]
```diff
@@ -156,6 +157,20 @@ public FontAttributes FontAttributes
             get => (FontAttributes)GetValue(FontAttributesProperty);
             set => SetValue(FontAttributesProperty, value);
         }
+        
+        public static BindableProperty IsRoundedProperty =
+	        BindableProperty.Create(nameof(IsRounded), typeof(bool), typeof(BadgeView), true,
+		        propertyChanged:IsRoundedChanged);
```
Comment (Cfun1) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:163]:
```suggestion
        /// <summary>
        /// Backing BindableProperty for the <see cref="IsRounded"/> property.
        /// </summary>
        public static BindableProperty IsRoundedProperty =
	        BindableProperty.Create(nameof(IsRounded), typeof(bool), typeof(BadgeView), true,
		        propertyChanged:IsRoundedChanged);
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 8]
```diff
@@ -234,7 +249,7 @@ void UpdateLayout()
             BadgeIndicatorBackground.BackgroundColor = BackgroundColor;
             BadgeIndicatorBackground.BorderColor = BorderColor;
             BadgeIndicatorBackground.HasShadow = HasShadow;
-
+
```
Comment (Cfun1) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:252]:
```suggestion

```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 9]
```diff
@@ -279,11 +300,28 @@ void UpdateBadgeViewPlacement(bool force = false)
...
+	        if (BadgeText.Text.Length == 1 && IsRounded)
+	        {
+		        if (Device.RuntimePlatform == Device.Android)
+		        {
+			        return new Thickness(10,0,10,0);
+		        }
+		        else
+		        {
+			        return new Thickness(8,0,8,0);
+		        }
```
Comment (Cfun1) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:320]:
```suggestion
		        if (Device.RuntimePlatform == Device.Android)
			        return new Thickness(10,0,10,0);
		        else
			        return new Thickness(8,0,8,0);
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 10]
```diff
@@ -354,14 +392,29 @@ async Task UpdateVisibilityAsync()
 
                 isVisible = badgeIsVisible;
             }
-            else
-                BadgeIndicatorContainer.IsVisible = badgeIsVisible;
+            else BadgeIndicatorContainer.IsVisible = badgeIsVisible;
```
Comment (Cfun1) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:395]:
```suggestion
            else
                BadgeIndicatorContainer.IsVisible = badgeIsVisible;
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 11]
```diff
@@ -354,14 +392,29 @@ async Task UpdateVisibilityAsync()
...
         }
-
+        
+        void OnBadgeIndicatorSizeChanged(object sender, EventArgs e)
+        {
+	        if (!IsRounded)
+	        {
+		        BadgeIndicatorBackground.CornerRadius = Device.RuntimePlatform == Device.Android ? 12 : 8;
+	        }
```
Comment (Cfun1) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:403]:
```suggestion
	        if (!IsRounded)
		        BadgeIndicatorBackground.CornerRadius = Device.RuntimePlatform == Device.Android ? 12 : 8;
```

File: src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs [hunk 12]
```diff
@@ -354,14 +392,29 @@ async Task UpdateVisibilityAsync()
...
+	        else
+	        {
+		        var min = Math.Min(BadgeIndicatorBackground.Width, BadgeIndicatorBackground.Height);
+		        if (min != -1)
+		        {
+			        BadgeIndicatorBackground.CornerRadius = (float) (min / 2);
+		        }
```
Comment (Cfun1) [inline src/CommunityToolkit/Xamarin.CommunityToolkit/Views/BadgeView/BadgeView.shared.cs:410]:
```suggestion
		        if (min != -1)
			        BadgeIndicatorBackground.CornerRadius = (float) (min / 2);
```

**Review by Cfun1 — CHANGES_REQUESTED** [review]

**Comment by jfversluis** [comment]
Hey @jmmortega sorry for getting back to this so late... Are you still interested in fixing this? :)

I think it makes sense to make the badges round, I'm not sure on the approach here though... Could we change it to open up the corner radius to people? We will make it round by default, but if they want anything different they can just use the corner radius to do something else?

If you're not interested anymore for whatever reason, I totally understand, just please let us know so we or someone else can take it. Thanks!

**Comment by Cfun1** [comment]
Trying to draw some attention to this pr, I guess it needs a re-target to develop.
@jfversluis, @pictos @AndreiMisiukevich

**Comment by AndreiMisiukevich** [comment]
@Cfun1 thank you :)

Yes please @jmmortega could you please retarget this PR to develop branch?

**Comment by brminnick** [comment]
@jmmortega Could you please resolve the merge conflicts? Thanks!

**Comment by eddie010** [comment]
Really looking forward to this getting implemented!

**Comment by AlleSchonWeg** [comment]
Any news here? Could somebody resolve the conflicts and merge?!
Thx.

**Comment by brminnick** [comment]
Closing PR as it has been abandoned.

If anybody is interested in taking over this PR we are more than happy to re-open it!

**Pull request closed without merging**